### PR TITLE
feat: check image checksum

### PR DESCRIPTION
**Pull request by SdgJlbl**
## Description

Reviewable commit by commit

Companion PR 
- https://github.com/Substra/substra-backend/pull/739
- https://github.com/Substra/orchestrator/pull/288
- https://github.com/Substra/substra/pull/393
- https://github.com/Substra/substrafl/pull/181

Fixes FL-1151, FL-1149, FL-1161

## TODO 
- [ ] check possible impact on frontend 
- [x] Failed builds are not propagating to compute task status (to check in orchestrator)

## How has this been tested?

Tested on a local deployment with Titanic (modified to run on 2 orgs)

## Checklist

- [ ] [changelog](../CHANGELOG.md) was updated with notable changes
- [ ] documentation was updated

### CODE REVIEW

**Review by oleobal — APPROVED** [review]
I might be lacking in context, so feel free to wait for further input, but this seems good

File: backend/orchestrator/function_pb2.pyi
Comment (thbcmlowk) [inline backend/orchestrator/function_pb2.pyi:169]: I see changes there but no companion PR linked to an update of the orchestrator. Am I missing something?

File: backend/api/tests/views/test_views_authentication.py
Comment (SdgJlbl) [reply backend/api/tests/views/test_views_authentication.py:42]: No tests at all on anything 🙃

Comment (thbcmlowk) [reply backend/api/tests/views/test_views_authentication.py:42]: No doubts then, perfect!

File: backend/orchestrator/function_pb2.pyi
Comment (thbcmlowk) [reply backend/orchestrator/function_pb2.pyi:169]: Thanks for the update - Or sorry if I missed the link in the first place 🙏

**Comment by thbcmlowk** [comment]
Can we not merge the PR until the orchestrator part is done (if not)?

**Comment by ThibaultFy** [comment]
/e2e --tests sdk,substrafl --refs orchestrator=feat/add-function-image,substra-backend=feat/check-image-checksum,substra=feat/rename-function-to-archive,substrafl=feat/rename-function-to-archive

**Comment by Owlfred** [comment]
End to end tests: :hourglass_flowing_sand: **STARTING**

**Comment by ThibaultFy** [comment]
/e2e --tests sdk,substrafl --refs orchestrator=feat/add-function-image,substra-backend=feat/check-image-checksum,substra=feat/rename-function-to-archive,substrafl=feat/rename-function-to-archive

**Comment by Owlfred** [comment]
End to end tests: :hourglass_flowing_sand: **STARTING**

File: backend/substrapp/compute_tasks/image_builder.py
Comment (guilhem-barthes) [inline backend/substrapp/compute_tasks/image_builder.py:56]: Some commented code to delete. Also, the comment suggests that we should use `api_function` isntead of `function` but the `.get` do otherwise

File: backend/substrapp/tasks/tasks_save_image.py
Comment (guilhem-barthes) [inline backend/substrapp/tasks/tasks_save_image.py:117]: As we use this function twice in this file, is it worth creating a function containing this logic?

**Review by guilhem-barthes — CHANGES_REQUESTED** [review]
Thanks for your work! Just some changes needed about some parsing seemingly done twice, some comments not matching the logic and commented code.

**Comment by ThibaultFy** [comment]
/e2e --tests sdk,substrafl --refs orchestrator=feat/add-function-image,substra-backend=feat/check-image-checksum,substra=feat/rename-function-to-archive,substrafl=feat/rename-function-to-archive

**Comment by Owlfred** [comment]
End to end tests: :hourglass_flowing_sand: **STARTING**

File: backend/substrapp/tasks/tasks_save_image.py
Comment (SdgJlbl) [inline backend/substrapp/tasks/tasks_save_image.py:63]: Is this an acceptable workaround? 🤔

**Comment by ThibaultFy** [comment]
/e2e --tests sdk,substrafl --refs orchestrator=feat/add-function-image,substra-backend=feat/check-image-checksum,substra=feat/rename-function-to-archive,substrafl=feat/rename-function-to-archive

**Comment by Owlfred** [comment]
End to end tests: :hourglass_flowing_sand: **STARTING**

**Comment by ThibaultFy** [comment]
The errors on the e2e tests might come from the orchestrator. When running locally, I get the following log:
```
│ {"level":"info","requestID":"47c3c1a9","msg":"Exec","level":"info","sql":"INSERT INTO addressables (storage_address,checksum) VALUES ($1,$2)","time":"2024-01-11T16:04:02Z","message":" │
│ panic: runtime error: invalid memory address or nil pointer dereference
```

**Comment by ThibaultFy** [comment]
/e2e --tests sdk,mnist --refs orchestrator=fix/new-function-have-no-image,substra-backend=feat/check-image-checksum,substra=feat/rename-function-to-archive,substrafl=feat/rename-function-to-archive

**Comment by Owlfred** [comment]
End to end tests: :hourglass_flowing_sand: **STARTING**

**Comment by ThibaultFy** [comment]
/e2e --tests sdk,substrafl --refs orchestrator=fix/new-function-have-no-image,substra-backend=feat/check-image-checksum,substra=feat/rename-function-to-archive,substrafl=feat/rename-function-to-archive

**Comment by Owlfred** [comment]
End to end tests: :hourglass_flowing_sand: **STARTING**

**Comment by ThibaultFy** [comment]
/e2e --tests sdk,substrafl,camelyon,doc --refs orchestrator=fix/new-function-have-no-image,substra-backend=feat/check-image-checksum,substra=feat/rename-function-to-archive,substrafl=feat/rename-function-to-archive

File: backend/substrapp/compute_tasks/image_builder.py
Comment (ThibaultFy) [reply backend/substrapp/compute_tasks/image_builder.py]: Back to api function in the end, for the reasons indicated in the comment

File: backend/api/events/sync.py
Comment (ThibaultFy) [reply backend/api/events/sync.py:96]: Format are not compatible if done on a one liner

**Comment by ThibaultFy** [comment]
/e2e --tests sdk,mnist,substrafl,camelyon --refs orchestrator=fix/new-function-have-no-image,substra-backend=feat/check-image-checksum,substra=feat/rename-function-to-archive,substrafl=feat/rename-function-to-archive

**Comment by Owlfred** [comment]
End to end tests: :hourglass_flowing_sand: **STARTING**

**Comment by ThibaultFy** [comment]
/e2e --tests sdk,mnist,substrafl,camelyon --refs orchestrator=feat/add-function-image,substra-backend=feat/check-image-checksum,substra=feat/rename-function-to-archive,substrafl=feat/rename-function-to-archive

**Comment by Owlfred** [comment]
End to end tests: :hourglass_flowing_sand: **STARTING**

File: backend/api/events/sync.py
Comment (guilhem-barthes) [reply backend/api/events/sync.py:96]: Giving a closer look, we do not use `from_grpc` anywhere else in this file. Perhaps we should just use the dict `function.get("image", None)` and adapt accordingly the `_update_function`

File: backend/substrapp/compute_tasks/image_builder.py
Comment (guilhem-barthes) [inline backend/substrapp/compute_tasks/image_builder.py:44]: This comment seems to either go higher (L43) or lower (before L50)

File: backend/substrapp/compute_tasks/image_builder.py
Comment (guilhem-barthes) [inline backend/substrapp/compute_tasks/image_builder.py:54]: Could we use `client.query_function(function.key)` instead of `ApiFunction.objects.get(key=function.key)` to avoid creating more dependencies between worker and DB?

File: backend/substrapp/tasks/tasks_save_image.py
Comment (guilhem-barthes) [reply backend/substrapp/tasks/tasks_save_image.py:65]: Why can't we make it optional? I would prefer not creating new links between workers and DB when it is not absolutely necessary

**Review by guilhem-barthes — COMMENTED** [review]
Thanks for your contribution. I feel some points need to be clarified before merging it in main

File: backend/substrapp/tasks/tasks_save_image.py
Comment (SdgJlbl) [reply backend/substrapp/tasks/tasks_save_image.py:65]: I agree it's not a long term solution, just wondering if we want to fix it now, or merge this before it gets bigger and fix as a follow-up 🤔

File: backend/substrapp/compute_tasks/image_builder.py
Comment (ThibaultFy) [reply backend/substrapp/compute_tasks/image_builder.py:54]: What client are you refering to ? Because from what I understand, this is to force the `organization_client` to get the updated value of the function.

**Comment by ThibaultFy** [comment]
/e2e --tests sdk --refs orchestrator=feat/add-function-image,substra-backend=feat/check-image-checksum,substra=feat/rename-function-to-archive,substrafl=feat/rename-function-to-archive

**Comment by ThibaultFy** [comment]
/e2e --tests sdk --refs orchestrator=feat/add-function-image,substra-backend=feat/check-image-checksum,substra=feat/rename-function-to-archive,substrafl=feat/rename-function-to-archive

**Comment by Owlfred** [comment]
End to end tests: :hourglass_flowing_sand: **STARTING**

**Comment by ThibaultFy** [comment]
/e2e --tests sdk,mnist,substrafl,camelyon --refs orchestrator=feat/add-function-image,substra-backend=feat/check-image-checksum,substra=feat/rename-function-to-archive,substrafl=feat/rename-function-to-archive

**Comment by Owlfred** [comment]
End to end tests: :hourglass_flowing_sand: **STARTING**

File: backend/api/events/sync.py
Comment (SdgJlbl) [reply backend/api/events/sync.py]: `from_grpc` is used heavily in `substrapp.events.sync`, whereas `api.events.sync` uses DRF serializers. 
I'm always a bit queasy about how we shall handle that on a case by case basis. 
If we are using directly a `.get` from the payload, we won't have any validation (neither from DRF validators nor pydantic) :/

File: backend/substrapp/tasks/tasks_save_image.py
Comment (SdgJlbl) [reply backend/substrapp/tasks/tasks_save_image.py:119]: This still feels like a quite hacky way of obtaining the URL, I will open a follow-up issue.

File: backend/substrapp/tasks/tasks_save_image.py
Comment (SdgJlbl) [reply backend/substrapp/tasks/tasks_save_image.py:65]: up, did you create the card?

File: backend/orchestrator/common_pb2_grpc.pyi
Comment (SdgJlbl) [inline backend/orchestrator/common_pb2_grpc.pyi:8]: Did we change the version of protobuf used or something?

File: backend/substrapp/tasks/tasks_save_image.py
Comment (ThibaultFy) [reply backend/substrapp/tasks/tasks_save_image.py:65]: Thanks for the reminder 👼 https://linear.app/owkin/issue/FL-1411/name-optional-on-updatefunction-in-orchestrator

**Review by SdgJlbl — COMMENTED** [review]
Typos :)

File: backend/orchestrator/common_pb2_grpc.pyi
Comment (ThibaultFy) [reply backend/orchestrator/common_pb2_grpc.pyi:8]: 🤷‍♂️  no idea

**Review by guilhem-barthes — COMMENTED** [review]
Thanks for your work, some small changes

**Comment by ThibaultFy** [comment]
/e2e --tests sdk,mnist,substrafl,camelyon --refs orchestrator=feat/add-function-image,substra-backend=feat/check-image-checksum,substra=feat/rename-function-to-archive,substrafl=feat/rename-function-to-archive

**Comment by Owlfred** [comment]
End to end tests: :hourglass_flowing_sand: **STARTING**

**Comment by ThibaultFy** [comment]
/e2e --tests sdk,mnist,substrafl,camelyon --refs orchestrator=feat/add-function-image,substra-backend=feat/check-image-checksum,substra=feat/rename-function-to-archive,substrafl=feat/rename-function-to-archive

**Comment by Owlfred** [comment]
End to end tests: :hourglass_flowing_sand: **STARTING**

**Review by guilhem-barthes — APPROVED** [review]
Thanks a lot 🏁

**Pull request merged**
